cli/coverage_cmd.go

```diff
@@ -0,0 +1,186 @@
+package cli
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+
+	"sourcegraph.com/sourcegraph/go-flags"
+
+	"sourcegraph.com/sourcegraph/srclib/config"
+	"sourcegraph.com/sourcegraph/srclib/graph"
+	"sourcegraph.com/sourcegraph/srclib/grapher"
+	"sourcegraph.com/sourcegraph/srclib/plan"
+)
+
+func init() {
+	cliInit = append(cliInit, func(cli *flags.Command) {
+		_, err := cli.AddCommand("coverage",
+			"srclib coverage",
+			"compute approximate amount of code successfully analyzed by srclib",
+			&coverageCmd,
+		)
+		if err != nil {
+			log.Fatal(err)
+		}
+	})
+}
+
+type Coverage struct {
+	FileScore      float32  // % files successfully processed
+	RefScore       float32  // % internal refs that resolve to a def
+	TokDensity     float32  // average number of refs/defs per LoC
+	UncoveredFiles []string // files for which srclib data was not successfully generated (best-effort guess)
+}
+
+type codeFileDatum struct {
+	LoC          int
+	NumRefs      int
+	NumDefs      int
+	NumRefsValid int
+}
+
+type CoverageCmd struct {
+}
+
+var coverageCmd CoverageCmd
+
+func (c *CoverageCmd) Execute(args []string) error {
+	repo, err := OpenLocalRepo()
+	if err != nil {
+		return err
+	}
+
+	cvg, err := coverage(repo)
+	if err != nil {
+		return err
+	}
+
+	out, err := json.MarshalIndent(cvg, "", "  ")
+	if err != nil {
+		return err
+	}
+	fmt.Println(string(out))
+
+	return nil
+}
+
+var codeExts = []string{".go", ".java", ".py", ".rb", ".cpp", ".ts", ".cs"} // codeExt lists all file extensions that indicate a code file we want to cover
+var codeExts_ = make(map[string]struct{})
+
+func init() {
+	for _, ext := range codeExts {
+		codeExts_[ext] = struct{}{}
+	}
+}
+
+func coverage(repo *Repo) (*Coverage, error) {
+	lineSep := []byte{'\n'}
+	codeFileData := make(map[string]*codeFileDatum)
+	log.Printf(repo.RootDir)
+	filepath.Walk(repo.RootDir, func(path string, info os.FileInfo, err error) error {
+		if filepath.IsAbs(path) {
+			var err error
+			path, err = filepath.Rel(repo.RootDir, path)
+			if err != nil {
+				return err
+			}
+		}
+
+		if info.IsDir() {
```

Comment:
It may also be worthwhile to check whether the directory name starts with `.` and if so return `filepath.SkipDir`.